### DISCUSSION TITLE
custom task loggers

### DISCUSSION THREAD

**Discussion by alexagr** [langroid/langroid#622] [Q&A] [answered]
I wanted to implement my own Task logger instead of the two defaults ones, stored in `self.logger` and `self.tsv_logger`.
I thought that it will be as simple as reimplementing `log_message()` method. However it turned out that it's a bit more tricky than that. 

Task's constructor initializes both members to None:
```
        self.logger: None | RichFileLogger = None
        self.tsv_logger: None | logging.Logger = None
```
But then there is `init()` method - that is called from `_pre_run_loop()` - that in turn is called from `run()` and `run_async()`.
And it has the following code:
```
        if self.caller is not None and self.caller.logger is not None:
            self.logger = self.caller.logger
        else:
            self.logger = RichFileLogger(
                str(Path(self.config.logs_dir) / f"{self.name}.log"),
                color=self.color_log,
            )
```
And the same for `self.tsv_logger`.

So not only it's impossible to leave these two members at `None`. It's also meaningless to assign them with any values after Task construction. As essentially everything will be "overwritten" by `init()` method.

I worked this around by reimplementing `init()` method in my Task class as follows:
```
    def init(
        self,
        msg: None | str | ChatDocument = None
    ) -> ChatDocument | None:
        # langroid always initializes logger and tsv_logger in init() method
        # so we need this dirty trick to avoid it
        self.caller = Task()
        self.caller.logger = NoOpRichFileLogger()
        self.caller.tsv_logger = NoOpLogger('dummy')

        response = super().init(msg)

        self.caller = None

        return response
```
But it's a pretty dirty trick, isn't it?

I wonder if we should add `if self.logger is None` to the `init()` code?
Or maybe there's another more elegant solution to this problem?

**Comment by alexagr** [comment]
And on a related note - there are quite a few places in `Agent` and `ChatAgent` classes that output data via `console.print()` or `print()`. The only way to disable them is to set `settings.quiet = True`. But that also disables `show_agent_response` and `show_llm_response` callbacks - and this is not suitable for "interactive" session, where responses need to be displayed to user - just not via regular console.

What would be the correct way to resolve this? Move calls to `show_agent_response` and `show_llm_response` callback out of `if not settings.quiet` block?

**Reply by pchalasani** [reply]
Are you looking for a granular way to selectively suppress some types of responses only, i.e. there are 3 responders `llm_response`, `agent_response`, and `user_response` -- and you're looking to suppress whichever combination you want?

**Reply by alexagr** [reply]
@pchalasani no - I'm talking about the following use-cases:

1. I write my own UI on top of langroid - similar to the chainlit integration that is in `agent/callbacks/chainlit.py`
2. I write REST/Websocket based API that provides "user input" to langroid-based engine - and returns "responses" back to the API  caller

In both cases I don't want console log to be "polluted" by print's because:
1. they may contain sensitive information
2. they are treated as "container/pod logs" by Docker / k8s

`settings.quiet` does exactly that - suppresses all print's. But I think that it SHOULD NOT suppress callbacks.

**Reply by pchalasani** [reply, marked as answer]
I see what you mean, your PR looks good!

**Comment by pchalasani** [comment]
> I wonder if we should add `if self.logger is None` to the `init()` code?

yes I think that would be reasonable. The original idea was to use the caller's logger if any. If a sub-task sets its own logger, then we'd have the logs split between the parent and child tasks, which is fine as long as the dev is aware of it.
